src/crimson/common/tri_mutex.cc

```diff
@@ -129,6 +130,7 @@ void tri_mutex::demote_to_write()
   assert(exclusively_used);
   exclusively_used = false;
   ++writers;
+  wake(type_t::write);
```

Comment:
Can we instead change `lock_state_t::demote` to be responsible for `unlock_for_excl()` and then calling `lock_for_write()` or `lock_for_read`?  and perhaps remove `demote_to_*` concept from tri_mutex?

Reply:
This should help with maintaining the current `wake` logic we have, unless I'm missing something else.

Reply:
> This should help with maintaining the current `wake` logic we have, unless I'm missing something else.

If this is all about keeping the the assertion in `wake()`, I think maybe we can change the code in the following way:
```
#ifndef NDEBUG
switch(type) {
case type_t::none:
  assert(!readers && !writers && !exclusively_used);
case type::read:
  assert(!writers && !exclusively_used);
case type::write:
  assert(!readers && !exclusively_used);
default:
  ceph_abort();
}
#endif
```
What do you think?

Reply:
@Matan-B Replacing unlock_for_excl() with unlock_for_excl(), lock_for_* would have different semantics -- there'd be no guarantee that it would actually get the second lock synchronously as another waiter might already be present.  The current main user requires that we be able to read/populate an obc and demote it to the correct lock type without allowing another waiter to get the lock -- it seems like this requires a special demote_to* concept.